[SPARK-5600] [core] Clean up FsHistoryProvider test, fix app sort order.

vanzin opened the pull request:
Clean up some test setup code to remove duplicate instantiation of the
provider. Also make sure unfinished apps are sorted correctly.

Commit:
[SPARK-5600] [core] Clean up FsHistoryProvider test, fix app sort order.

Clean up some test setup code to remove duplicate instantiation of the
provider. Also make sure unfinished apps are sorted correctly.

SparkQA commented:
[Test build #26768 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26768/consoleFull) for   PR 4370 at commit [`298371c`](https://github.com/apache/spark/commit/298371c46b053af0235ded7704179b4359e16f47).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26768 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26768/consoleFull) for   PR 4370 at commit [`298371c`](https://github.com/apache/spark/commit/298371c46b053af0235ded7704179b4359e16f47).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26768/
Test FAILed.

vanzin commented:
Jenkins, retest this please.

SparkQA commented:
[Test build #26779 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26779/consoleFull) for   PR 4370 at commit [`298371c`](https://github.com/apache/spark/commit/298371c46b053af0235ded7704179b4359e16f47).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26779 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26779/consoleFull) for   PR 4370 at commit [`298371c`](https://github.com/apache/spark/commit/298371c46b053af0235ded7704179b4359e16f47).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26779/
Test PASSed.

srowen commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -230,6 +230,17 @@ private[history] class FsHistoryProvider(conf: SparkConf) extends ApplicationHis
+  private def compareAppInfo(
+      i1: FsApplicationHistoryInfo,
+      i2: FsApplicationHistoryInfo): Boolean = {
+    i1.endTime >= i2.endTime && i1.startTime >= i2.startTime
```
The rest LGTM, but, isn't this defining a different and inconsistent sort order? For example if the first argument's start and end time are after and before the other's start and times, respectively, then the method returns false both ways, when one has to be true. The existing logic is to sort by endTime descending, then startTime descending, so I think it's more like...

```
if (i1.endTime != i2.endTime) {
  i1.endTime > i2.endTime
} else {
  i1.startTime > i2.startTime
}
```

srowen commented on core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala:
```diff
@@ -211,4 +202,8 @@ class FsHistoryProviderSuite extends FunSuite with BeforeAndAfter with Matchers
     new FileOutputStream(file).close()
   }
 
+  private def createTestConf(): SparkConf = {
```
Could/should this be method be reused in more places now that your other change to this test is merged?

vanzin replied:
Yes, I'll merge with master and update the PR.

Commit:
Merge branch 'master' into SPARK-5600

Commit:
Merge branch 'master' into SPARK-5600

Commit:
Review feedback.

Commit:
Cleanup test code a bit.

Rename variables / apps so that it's easier to follow which one
is which, and also tweak the app times so that they better test
the sorting code.

SparkQA commented:
[Test build #26939 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26939/consoleFull) for   PR 4370 at commit [`0d048d5`](https://github.com/apache/spark/commit/0d048d5f3dfd2f71c6060b3cc53bc332df14fa2e).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26939 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26939/consoleFull) for   PR 4370 at commit [`0d048d5`](https://github.com/apache/spark/commit/0d048d5f3dfd2f71c6060b3cc53bc332df14fa2e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `      "public class " + className + extendsText + " implements java.io.Serializable `
  * `  case class RegisterExecutor(`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26939/
Test PASSed.

andrewor14 commented:
LGTM I'm merging this into master 1.3 thanks @vanzin

srowen commented on core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:
```diff
@@ -231,6 +231,17 @@ private[history] class FsHistoryProvider(conf: SparkConf) extends ApplicationHis
+  private def compareAppInfo(
+      i1: FsApplicationHistoryInfo,
+      i2: FsApplicationHistoryInfo): Boolean = {
+    if (i1.endTime != i2.endTime) i1.endTime >= i2.endTime else i1.startTime >= i2.startTime
```
I don't imagine it probably matters, but this should be `>` instead of `>=` I think? or else this says that two equal intervals should both precede the other. I bet it doesn't make the sort wrong, just might affect whether the sort is stable, but who cares here.
